### PR TITLE
tests: enviroment.py: add prplWRT entities

### PR DESCRIPTION
**Pull request by abelog**
## Description
This PR introduces basic prpMesh-specific entities according to [UML](https://raw.githubusercontent.com/prplfoundation/prplMesh/master/documentation/images/plantuml/boardfarm_prplMesh_classes.png) and is a part of #1072 task.
As yet there is no config, and introduced classes are not yet used - code is totally untested.
Fixes are expected in separate PR, after example config and documentation about it creation is merged (it also will be done in separate PR).
Goal of this PR is to bring prplWRT entities with enough functionality to pass `InitialApConfig`

## Changes
Add:
- Class `RadioHostapd` 
- Class `ALEntityPrplWrt`
-  Stub class `VirtualAPHostapd` (more functionality isn't required for `InitialApConfig` test)

### CODE REVIEW

File: tests/environment.py
Comment (arnout) [inline tests/environment.py:218]: This assumes that the log line we're looking for appears in the last 10 lines. I think we should get the entire file instead. And we can actually make use of start_line here.
```suggestion
    device.sendline("tail -n +{:d} -f {}".format(start_line + 1, log_path))
```
(note that line numbering is 1-based, not 0-based, hence the +1)

File: tests/environment.py
Comment (arnout) [inline tests/environment.py:215]: `None` can't be right, it's not a type. If you really expect device to be `None`, it should be `NoneType`. But that can't be right :-) It should actually be PrplMeshPrplWRT, but then you create a circular dependency... So, with nothing better, just drop the type specification.

However, I think it would be better to make this function a member of PrplMeshPrplWRT itself. Also, I think everything defined in this commit should move to prplmesh_prplwrt.py (which, BTW, nicely skates around the circular dependency issue).

File: tests/environment.py
Comment (arnout) [inline tests/environment.py:439]: No need for a `device` parameter here - you can use `self.agent.device`.

File: tests/environment.py
Comment (arnout) [inline tests/environment.py:397]: I believe we can get this out of the config file as well. Hardcoding it is bad.

File: tests/environment.py
Comment (arnout) [inline tests/environment.py:405]: Personally I'd prefer to do the parsing in Python instead of shell, but that's just me I guess.
```suggestion
        ucc_port_string = self.command('grep "ucc_listener_port" {}".format(self.config_file_name))
        ucc_port, _ = ucc_port_string.split(' = ', 1)
```

File: tests/environment.py
Comment (RanRegev) [inline tests/environment.py:217]: nitpick 
```suggestion
    """Waits for log matching regex expression to show up."""
```

File: tests/environment.py
Comment (RanRegev) [inline tests/environment.py:218]: use long format in scripts. It increases readability and requires less efforts from a newcomer
```suggestion
    device.sendline("tail --lines=10 --follow {}".format(log_path))
```

File: tests/environment.py
Comment (RanRegev) [inline tests/environment.py:394]: Creating an object with an embedded invariant is a bad idea. The source of the problem is in the base class `ALEntity` that behaves _as if_ it is a general entity but in reality has to know if it a controller or agent.
However, we don't have to follow this guideline, nor we have to fix it in the base.
All that is required is to create two classes: one for `controller` and the second for `agent`.
See the comment above.

File: tests/environment.py
Comment (RanRegev) [inline tests/environment.py:391]:
```suggestion
class ALEntityPrplWrtController(ALEntity):
# and later: class ALEntityPrplWrtAgent(ALEntity):
```
See detailed comment below.

File: tests/environment.py
Comment (arnout) [inline tests/environment.py:408]: `head` is redundant since you do regex search below.

File: tests/environment.py
Comment (RanRegev) [inline tests/environment.py:405]:
```suggestion
                                "| cut --delimiter='=' --fields=2 | cut --delimiter\" \" -fields=1").format(self.config_file_name))
```

File: tests/environment.py
Comment (arnout) [inline tests/environment.py:440]: If it's the interface *name* call it iface_name, not bssid.

If it is the radio MAC address, call is radio_mac or radio_uid, not bssid.

bssid is definitely incorrect here.

File: tests/environment.py
Comment (RanRegev) [inline tests/environment.py:404]: These kind of commands are _extremely_ hard to maintain.
One important helper is a comment above this command with the expected `grep` output.
After that, to be even more helpful, adding an explanation of how the ucc_port is extracted out of the expected result using the piped commands - would be even better.
Below there is more code that this comment should be applied to. I **didn't** add it to each line.

File: tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py
Comment (arnout) [inline tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:55]: (nitpick)
```suggestion
        return match == 1
```

However, I thought the returned index was 0-based, so shouldn't this be
```suggestion
        return match == 0
```
?

File: tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py
Comment (arnout) [inline tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:34]: This can't be right, the device argument is missing... I guess this should be
```suggestion
            self.controller_entity = ALEntityPrplWrt(self.name, self, is_controller=True)
```
In that case, there is no need to pass self.name, right? The __init__ function can just get it from the device argument.

**Review by arnout — APPROVED** [review]
As usual, lots of small comments. It's architecturally sound so can be merged after fixing up the review comments.

File: tests/environment.py
Comment (RanRegev) [inline tests/environment.py:410]: Regular expressions are better be explained explicitly. They are not part of the `common` or `expected` knowledge from the average maintainer. Moreover, there are few dialects of regexp so even if one is familiar with regexp she would still need to look for the specific one (python in this case).

File: tests/environment.py
Comment (RanRegev) [inline tests/environment.py:428]: I searched for the means that I have, as a user of this function, to supply a value for `start_line` - and i couldn't find.
I don't think that there is a way for the user to get a value for it. I also looked at [test_flow.py](https://github.com/prplfoundation/prplMesh/blob/feature/n1072-prplwrt-device/tests/test_flows.py) and couldn't find a use of it.
Normally the interface exposes a way to know where to start the next search, e.g. by returning the last position it was found, or by explicit function call `get_last_searched_line()`

File: tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py
Comment (RanRegev) [inline tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:69]: Can you elaborate in the comments how `pass` prevents the disconnection?

**Review by RanRegev — APPROVED** [review]

File: tests/environment.py
Comment (arnout) [reply tests/environment.py:218]: Unfortunately, [long options are non-POSIX](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/tail.html#tag_20_125_04) so we can't be sure that the `tail` implementation on target supports them. I believe in the busybox implementation, the long option support is optional.

File: tests/environment.py
Comment (arnout) [reply tests/environment.py:394]: > All that is required is to create two classes: one for controller and the second for agent.

The problem with that is that we would need to create 6 additional classes, and we would have to use multiple inheritance or duplicate code.

File: tests/environment.py
Comment (arnout) [reply tests/environment.py:405]: Again, [not POSIX](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/cut.html)

File: tests/environment.py
Comment (RanRegev) [reply tests/environment.py:394]: * Duplicated code is a bad sign, so we'll avoid this.
* 6 additional classes does not imply wrong.
* multiple inheritance is a design decision

We _have_ variants, otherwise `is_controller` wouldn't be introduced.
The question is how to _express_ them.
I didn't look deeply into _other implementation options_ just indicated that `is_controller` is a red flag that complicates the code. I think we should implement this variant in a different way.

File: tests/environment.py
Comment (arnout) [reply tests/environment.py:428]: > Normally the interface exposes a way to know where to start the next search, e.g. by returning the last position it was found

[That's exactly what it does.](https://github.com/prplfoundation/prplMesh/blob/master/tests/environment.py#L187) However, this was never documented because I consider it a horrible interface, and it should be converted to use a [checkpoint mechanism](https://github.com/prplfoundation/prplMesh/blob/master/tests/environment.py#L155).

File: tests/environment.py
Comment (abelog) [reply tests/environment.py:218]: Fixed in #1358. `start_line` now counted in.

File: tests/environment.py
Comment (abelog) [reply tests/environment.py:397]: Fixed in #1358. Bridge is `grep`-ed from config file.

File: tests/environment.py
Comment (abelog) [reply tests/environment.py:405]: Parsing moved to Python primary. Fixed in #1358.

File: tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py
Comment (abelog) [reply tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:55]: Yes, you are right. Fixed in #1358.

File: tests/environment.py
Comment (abelog) [reply tests/environment.py:218]: Sadly, but `tail` of `busybox` used in RAX40 images doesn't support `--lines` option. But, i was checking pre-RC1 image of prplWRT on RAX40.

File: tests/environment.py
Comment (abelog) [reply tests/environment.py:439]: Yes. It's totally redundant.

File: tests/environment.py
Comment (abelog) [reply tests/environment.py:440]: Fixed in #1358.

File: tests/environment.py
Comment (abelog) [reply tests/environment.py:408]: `head` is just to get last of IPs, since `ip` may report multiple IPs for same interface. E.g :
```
root@OpenWrt:/# ip -family inet addr show br-lan
13: br-lan: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP qlen 1000
    inet 192.168.1.1/24 brd 192.168.1.255 scope global br-lan
       valid_lft forever preferred_lft forever
    inet 192.168.250.170/24 brd 192.168.250.255 scope global br-lan
       valid_lft forever preferred_lft forever
    inet 172.21.0.245/16 scope global br-lan
       valid_lft forever preferred_lft forever
```

File: tests/environment.py
Comment (abelog) [reply tests/environment.py:217]: Fixed in #1358.

File: tests/environment.py
Comment (abelog) [reply tests/environment.py:218]: Resolving, as there is no way for us to use long options, as they are not supported on RAX40.

File: tests/environment.py
Comment (abelog) [reply tests/environment.py:405]: Resolving, as those options are not supported on RAX40.

File: tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py
Comment (abelog) [reply tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:69]: We currently are using UART, which do not have auto-hangup. Hence, no need in implementation of this method.

File: tests/environment.py
Comment (abelog) [reply tests/environment.py:439]: Resolved in #1358

File: tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py
Comment (abelog) [reply tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_prplwrt.py:34]: Than better to drop `name` argument at all. Fixed in #1358.

File: tests/environment.py
Comment (arnout) [reply tests/environment.py:408]: I'll repeat my statement: since you do a regex search below, head is redundant. re.search will return the first match, so whether you do the `head` or not, you get the same result.

**Comment by arnout** [comment]
@mergifyio rebase

**Pull request merged**
